sw/device/lib/base/ibex.c

```diff
@@ -26,12 +28,26 @@ status_t ibex_wait_rnd_valid(void) {
   }
 }
 
+status_t ibex_rnd_data_read(uint32_t *rnd_data) {
+  *rnd_data = abs_mmio_read32(kBase + RV_CORE_IBEX_RND_DATA_REG_OFFSET);
+  return OK_STATUS();
+}
+
 status_t ibex_rnd_status_read(uint32_t *rnd_status) {
   *rnd_status = abs_mmio_read32(kBase + RV_CORE_IBEX_RND_STATUS_REG_OFFSET);
   return OK_STATUS();
 }
 
-status_t ibex_rnd_data_read(uint32_t *rnd_data) {
-  *rnd_data = abs_mmio_read32(kBase + RV_CORE_IBEX_RND_DATA_REG_OFFSET);
-  return OK_STATUS();
+uint32_t ibex_rnd_uint32(void) {
+  uint32_t rnd_data;
+  status_t status = entropy_complex_check();
+
+  if (status_ok(status)) {
+    ibex_wait_rnd_valid();
+    ibex_rnd_data_read(&rnd_data);
+  } else {
+    rnd_data = (uint32_t)ibex_mcycle_read();
+  }
```

Comment:
I think this is very dangerous. A single interface that might return real randomness but also might return completely non-random values (cycle counts are _very_ weak randomness, especially because the high bits are likely to repeat themselves for long periods of time) leaves a lot of opportunity for callers to expect good randomness and get something that offers no protection. For example, a caller might reasonably call `ibex_rnd_uint32()` and then use the low 16 bits for one thing and the high 16 bits for some other smaller-than-32-bit value, without realizing the high bits will be the same for 2^16 cycles at a time. Poor-quality randomness should be returned in a separate interface.